src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java

```diff
@@ -22,32 +20,46 @@ public AnimalWebController(AnimalService animalService) {
 		this.animalService = animalService;
 	}
 
-	@GetMapping("")
-	public ModelAndView addAnimalView() {
-		ModelAndView modelAndView = new ModelAndView("addanimal");
-		modelAndView.addObject("animals", animalService.getAnimals());
-		modelAndView.addObject("animal", new Animal());
+	@GetMapping("/list")
+	public ModelAndView list() {
+		ModelAndView modelAndView = new ModelAndView("animal/animal_page");
+		List<Animal> animals = animalService.getAnimals();
+		modelAndView.addObject("animals", animals);
 		return modelAndView;
 	}
 
-	@PostMapping("")
-	public ModelAndView addAnimal(@ModelAttribute @Valid Animal animal) {
-		ModelAndView modelAndView = new ModelAndView();
-		try {
-			animalService.addAnimal(animal);
-		} catch (InvalidAnimalException e) {
-			e.printStackTrace();
-		}
-		modelAndView.addObject("animals", animalService.getAnimals());
-		modelAndView.setViewName("animals");
-		modelAndView.addObject("animal", new Animal());
+	@GetMapping("/add")
+	public ModelAndView add() {
+		ModelAndView modelAndView = new ModelAndView("animal/animal_form");
+		Animal animal = new Animal();
+		modelAndView.addObject("animal", animal);
 		return modelAndView;
 	}
 
-	@GetMapping("/showList")
-	public ModelAndView getAnimalsView() {
-		ModelAndView modelAndView = new ModelAndView("animals");
-		modelAndView.addObject("animals", animalService.getAnimals());
+	@GetMapping("/update/{id}")
+	public ModelAndView update(@PathVariable("id") Long id) throws AnimalNotFoundException {
+		ModelAndView modelAndView = new ModelAndView("animal/animal_form");
+		Animal animal = animalService.getAnimalById(id);
+		modelAndView.addObject("animal", animal);
 		return modelAndView;
 	}
+
+	@PostMapping("/save")
+	public ModelAndView save(@ModelAttribute("animal") Animal animal)
+			throws AnimalNotFoundException, InvalidAnimalException {
+		if (animal != null && animal.getId() != null) {
```

Comment:
if `animal == null` then you still call `addAnimal`, it is probably not correct, right?

Reply:
/add and /update both will lead to save, and save then decides if we are adding a new entity or updating it upon the animal id field.
The "if (animal!=null)" was just an extra, and useless, check. So I removed it.
I'm interested if you have a more proper way to handle the mvc crud operations.